datafusion/physical-expr/src/expressions/in_list.rs

```diff
@@ -70,320 +51,172 @@ impl Debug for InListExpr {
-            .collect::<Vec<_>>();
-
-        Ok(collection_contains_check!(array, values, $NEGATED, contains_null))
-    }};
-}
-
-macro_rules! set_contains_for_float {
-    ($ARRAY:expr, $SET_VALUES:expr, $SCALAR_VALUE:ident, $NEGATED:expr) => {{
-        let contains_null = $SET_VALUES.iter().any(|s| s.is_null());
-        let bool_array = if $NEGATED {
-            // Not in
-            if contains_null {
-                $ARRAY
-                    .iter()
-                    .map(|vop| {
-                        match vop.map(|v| !$SET_VALUES.contains(&v.try_into().unwrap())) {
-                            Some(true) => None,
-                            x => x,
-                        }
-                    })
-                    .collect::<BooleanArray>()
-            } else {
-                $ARRAY
-                    .iter()
-                    .map(|vop| vop.map(|v| !$SET_VALUES.contains(&v.try_into().unwrap())))
-                    .collect::<BooleanArray>()
-            }
-        } else {
-            // In
-            if contains_null {
-                $ARRAY
-                    .iter()
-                    .map(|vop| {
-                        match vop.map(|v| $SET_VALUES.contains(&v.try_into().unwrap())) {
-                            Some(false) => None,
-                            x => x,
-                        }
-                    })
-                    .collect::<BooleanArray>()
-            } else {
-                $ARRAY
-                    .iter()
-                    .map(|vop| vop.map(|v| $SET_VALUES.contains(&v.try_into().unwrap())))
-                    .collect::<BooleanArray>()
-            }
-        };
-        ColumnarValue::Array(Arc::new(bool_array))
-    }};
+impl<T> ArraySet<T>
+where
+    T: Array + From<ArrayData>,
+{
+    fn new(array: &T, hash_set: ArrayHashSet) -> Self {
+        Self {
+            array: T::from(array.data().clone()),
+            hash_set,
+        }
+    }
 }
 
-macro_rules! set_contains_for_primitive {
-    ($ARRAY:expr, $SET_VALUES:expr, $SCALAR_VALUE:ident, $NEGATED:expr) => {{
-        let contains_null = $SET_VALUES.iter().any(|s| s.is_null());
-        let native_set = $SET_VALUES
-            .iter()
-            .flat_map(|v| match v {
-                $SCALAR_VALUE(value, ..) => *value,
-                datatype => {
-                    unreachable!(
-                        "InList can't reach other data type {} for {}.",
-                        datatype, v
-                    )
-                }
+impl<T> Set for ArraySet<T>
+where
+    T: Array + 'static,
+    for<'a> &'a T: ArrayAccessor,
+    for<'a> <&'a T as ArrayAccessor>::Item: PartialEq + HashValue,
+{
+    fn contains(&self, v: &dyn Array, negated: bool) -> BooleanArray {
+        let v = v.as_any().downcast_ref::<T>().unwrap();
+        let in_data = self.array.data();
+        let in_array = &self.array;
+        let has_nulls = in_data.null_count() != 0;
+
+        ArrayIter::new(v)
```

Comment:
I wonder if it would be more performant to use `hash_array` to has the array all in one go rather than doing it an element at a time? 

I was thinking that since this is the hot path the more vectorization we get the better